lib/ansible/modules/cloud/amazon/iam_policy.py

```diff
@@ -112,240 +114,215 @@
-        error_msg = boto_exception(err)
-        module.fail_json(changed=changed, msg=error_msg)
+        if self.policy_name not in matching_policies or not self.skip_duplicates:
+            self.put(policy_doc)
+
+    def run(self):
+        if self.state == 'present':
+            self.create()
+        elif self.state == 'absent':
+            self.delete()
+        return {
+            'changed': self.changed,
+            self._iam_type() + '_name': self.name,
+            'policies': self.list()
+        }
+
+
+class UserPolicy(Policy):
+
+    @staticmethod
+    def _iam_type():
+        return 'user'
+
+    def _list(self, name):
+        return self.client.list_user_policies(UserName=name)
 
-    return changed, name, updated_policies, msg
+    def _get(self, name, policy_name):
+        return self.client.get_user_policy(UserName=name, PolicyName=policy_name)
+
+    def _put(self, name, policy_name, policy_doc):
+        return self.client.put_user_policy(UserName=name, PolicyName=policy_name, PolicyDocument=policy_doc)
+
+    def _delete(self, name, policy_name):
+        return self.client.delete_user_policy(UserName=name, PolicyName=policy_name)
+
+
+class RolePolicy(Policy):
+
+    @staticmethod
+    def _iam_type():
+        return 'role'
+
+    def _list(self, name):
+        return self.client.list_role_policies(RoleName=name)
+
+    def _get(self, name, policy_name):
+        return self.client.get_role_policy(RoleName=name, PolicyName=policy_name)
+
+    def _put(self, name, policy_name, policy_doc):
+        return self.client.put_role_policy(RoleName=name, PolicyName=policy_name, PolicyDocument=policy_doc)
+
+    def _delete(self, name, policy_name):
+        return self.client.delete_role_policy(RoleName=name, PolicyName=policy_name)
+
+
+class GroupPolicy(Policy):
+
+    @staticmethod
+    def _iam_type():
+        return 'group'
+
+    def _list(self, name):
+        return self.client.list_group_policies(GroupName=name)
+
+    def _get(self, name, policy_name):
+        return self.client.get_group_policy(GroupName=name, PolicyName=policy_name)
+
+    def _put(self, name, policy_name, policy_doc):
+        return self.client.put_group_policy(GroupName=name, PolicyName=policy_name, PolicyDocument=policy_doc)
+
+    def _delete(self, name, policy_name):
+        return self.client.delete_group_policy(GroupName=name, PolicyName=policy_name)
 
 
 def main():
-    argument_spec = ec2_argument_spec()
-    argument_spec.update(dict(
-        iam_type=dict(
-            default=None, required=True, choices=['user', 'group', 'role']),
-        state=dict(
-            default=None, required=True, choices=['present', 'absent']),
+    argument_spec = dict(
+        iam_type=dict(default=None, required=True, choices=['user', 'group', 'role']),
+        state=dict(default='present', choices=['present', 'absent']),
         iam_name=dict(default=None, required=False),
         policy_name=dict(default=None, required=True),
```

Comment:
Please remove this default too
```suggestion
        policy_name=dict(required=True),
```